plugin.video.themoviedb.helper/resources/lib/player.py

```diff
@@ -0,0 +1,393 @@
+                        player = (resolve_url, folder[idx].get('file'))  # Set the folder path to open/play
+                        break  # Move onto next action
+                    else:  # Ask user to select a different player if no items in dialog
+                        return self.play_external(force_dialog=force_dialog, playerindex=playerindex)
+
+                x = 0
+                utils.kodi_log('Player -- Retrieved Folder\n{}'.format(string_format_map(player[1], self.item)), 2)
+                for f in folder:  # Iterate through plugin folder looking for a matching item
+                    x += 1  # Keep an index for position matching
+                    for k, v in d.items():  # Iterate through our key (infolabel) / value (infolabel must match) pairs of our action
+                        if k == 'position':  # We're looking for an item position not an infolabel
+                            if utils.try_parse_int(string_format_map(v, self.item)) != x:  # Format our position value
+                                break  # Not the item position we want so let's go to next item in folder
+                        elif not f.get(k) or not re.match(string_format_map(v, self.item), u'{}'.format(f.get(k, ''))):  # Format our value and check if it regex matches the infolabel key
+                            break  # Item's key value doesn't match value we are looking for so let's got to next item in folder
+                    else:  # Item matched our criteria so let's open it up
+                        utils.kodi_log('Player -- Found Match!\n{}'.format(f), 2)
+                        resolve_url = True if f.get('filetype') == 'file' else False  # Set true for files so we can play
+                        player = (resolve_url, f.get('file'))  # Get ListItem.FolderPath for item
+                        break  # Move onto next action (either open next folder or play file)
+                else:
+                    utils.kodi_log('Player -- Failed to find match!\n{}'.format(d), 2)
+                    return self.play_external(force_dialog=force_dialog, playerindex=playerindex)  # Ask user to select a different player
+
+        # Play/Search found item
+        if player and player[1]:
+            action = string_format_map(player[1], self.item)
+            if player[0] and action.endswith('.strm'):  # Action is play and is a strm so PlayMedia
+                utils.kodi_log(u'Player -- Found strm.\nAttempting PLAYMEDIA({})'.format(action), 2)
+                xbmc.executebuiltin(utils.try_decode_string(u'PlayMedia({0})'.format(action)))
+            elif player[0]:  # Action is play and not a strm so play with player
+                utils.kodi_log(u'Player -- Found file.\nAttempting to PLAY: {}'.format(action), 2)
+                xbmc.Player().play(action, ListItem(library='video', **self.details).set_listitem())
```

Comment:
plugins should not use xbmcPlayer to request playback. What is this for?

Reply:
I may have asked before

Reply:
TMDbHelper doesn't provide any playable content. It has user definable "player" files which provides rules about how to interact with other addons that do provide content. When a user clicks on an item, they are presented with a dialog to choose which "player" to use. JSONRPC is then used to walk a defined plugin path according to a set of rules. When a matching item is found, that plugin path is passed to Player() and combined with the detailed information that TMDbHelper has pulled from online APIs.

There's more detail in the wiki:
https://github.com/jurialmunkey/plugin.video.themoviedb.helper/wiki/Player-Function